🏗️ Build spider: Los Angeles City Planning

msrezaie opened the pull request:
## What's this PR do?
<!-- eg. This PR updates the scraper for Cleveland City Council because of changes to how they display their meeting schedule. -->
Adds a spider to scrape meeting information of Los Angeles City Planning. The scraper uses API requests to get the meeting items.

## Why are we doing this?
<!-- eg. The website's layout was recently updated, causing our existing scraper to fail. This change ensures our scraper remains functional and continues to provide timely updates on council meetings. -->
Scraper requested from [spreadsheet](https://docs.google.com/spreadsheets/d/1-HwPQKwm56vZ8h7CEM65GazKLTS0086gmaEzEHXp7O4/edit?usp=sharing).

## Steps to manually test
1. Ensure the project is installed:
```
pipenv sync --dev
```
2. Activate the virtual env and enter the pipenv shell:
```
pipenv shell
```
3. Run the spider:
```
scrapy crawl losca_City_Planning -O test_output.csv
```
4. Monitor the output and ensure no errors are raised.
5. Inspect `test_output.csv` to ensure the data looks valid.
6. Ensure all tests pass
```
pytest
```

## Are there any smells or added technical debt to note?
No

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced a new spider for scraping meeting data from the Los Angeles City Planning website.
	- Enhanced data validation and parsing for meeting entries.
	- Added a new JSON file containing detailed information about city planning hearings.

- **Tests**
	- Added a comprehensive suite of unit tests for the new LoscaCityPlanningSpider to ensure accurate data parsing and validation.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

msrezaie committed:
spider for Los Angeles City Planning

msrezaie committed:
fixed spider name

msrezaie committed:
format fix

msrezaie committed:
flake8 fix

msrezaie committed:
Update test_losca_City_Planning.py

LienDang commented on city_scrapers/spiders/losca_City_Planning.py:
```diff
@@ -0,0 +1,101 @@
+        f"https://planning.lacity.gov/dcpapi/meetings/api/all/commissions/{start_year}",
+        f"https://planning.lacity.gov/dcpapi/meetings/api/all/boards/{start_year}",
+        f"https://planning.lacity.gov/dcpapi/meetings/api/all/hearings/{start_year}",
+    ]
```
When I ran the test command, I realized it got less meetings than what are shown on the [source website](https://planning.lacity.gov/about/commissions-boards-hearings), eg. last meeting of the year 2024 [Area Planning Commission - 12/26/2024](https://planning.lacity.gov/dcpapi2/meetings/document/77972) is missing. It turned out that the website is using API v2 endpoints, while you're using API v1. 
So please update `start_urls` to API v2 endpoints:
```python
start_urls = [
    f"https://planning.lacity.gov/dcpapi2/meetings/api/all/commissions/{start_year}",
    f"https://planning.lacity.gov/dcpapi2/meetings/api/all/boards/{start_year}",
    f"https://planning.lacity.gov/dcpapi2/meetings/api/all/hearings/{start_year}",
]
```

Please see my test outputs for each `start_url` API v2 endpoint:
[test_output_all_hearings_v2.csv](https://github.com/user-attachments/files/18243454/test_output_all_hearings_v2.csv)
[test_output_boards_v2.csv](https://github.com/user-attachments/files/18243455/test_output_boards_v2.csv)
[test_output_commissions_v2.csv](https://github.com/user-attachments/files/18243457/test_output_commissions_v2.csv)

❗ One thing I noticed with the API v2 endpoints is that they seem to have rate limit. To produce those 3 test outputs files, I couldn't run the test commands continuously. Sometimes I got zero-byte output file, so I had to wait for about 1-2 minutes to rerun the command, and it was successful eventually. So when you update the API v2 endpoints, please make sure to take care of this issue.

LienDang commented on city_scrapers/spiders/losca_City_Planning.py:
```diff
@@ -0,0 +1,101 @@
+
+    def _parse_source(self, response):
+        """Parse or generate source."""
+        return response.url
```
We prefer website URL over API URL for `source` field. In this case it can be a hardcoded value `https://planning.lacity.gov/about/commissions-boards-hearings`

LienDang commented on city_scrapers/spiders/losca_City_Planning.py:
```diff
@@ -0,0 +1,101 @@
+        if not url:
+            return False
+        if not url.startswith("https://"):
+            self.logger.warning(f"Non-HTTPS URL found: {url}")
```
All the links returned from the API are "http". But when you visit them, they will be redirected to "https" protocol. So your validation here for https URLs is too strict, which results in no links are scraped. Please lessen the validation to accept both "http" and "https" URLs.

LienDang commented on city_scrapers/spiders/losca_City_Planning.py:
```diff
@@ -0,0 +1,101 @@
+            return False
+        return True
+
+    def _parse_links(self, item):
```
More links can be scraped here. Beside of "AgendaLink" and "AddDocsLink", there are "MinutesLink" and "AudioLink". Maybe you can loop through any key that ends with "Link" in the response, so "HearingLink" is possibly available.

Please see the sample response
```json
{
  "Date": "9/24/2024",
  "Type": "Area Planning Commission",
  "APC": "Central",
  "CaseNumber": "",
  "BoardName": "",
  "Note": "Central Meeting Agenda",
  "Address": "3300 West 8th Street",
  "CD": "",
  "Agenda": "77458",
  "AgendaLink": "http://planning.lacity.gov/dcpapi2/meetings/document/77458",
  "Audio": "77506",
  "AudioLink": "http://planning.lacity.gov/dcpapi2/meetings/document/77506",
  "Minutes": "77995",
  "MinutesLink": "http://planning.lacity.gov/dcpapi2/meetings/document/77995",
  "Hearing": "",
  "HearingLink": "",
  "AddDocs": "68122",
  "AddDocsLink": "http://planning.lacity.gov/dcpapi2/meetings/document/addtldoc/68122",
  "Time": "",
  "Description": "",
  "workshop": ""
}
```

LienDang commented on city_scrapers/spiders/losca_City_Planning.py:
```diff
@@ -0,0 +1,101 @@
+        return {
+            "address": address,
+            "name": item.get("BoardName", ""),
+        }
```
+ In this sample response, your regex pattern accidentally strips the first digit character "2" of the address.
```json
// current location output according to the regex pattern
{
  "address": "1101 West Ventura Boulevard, 1515 South Veteran Avenue, 1523 South Veteran Avenue, 1666 North Vermont Avenue, 1642 – 1666 North Vermont Avenue, 4646 – 4650 West Prospect Avenue, 4685 – 4697 West Hollywood Boulevard",
  "name": ""
}
```
+ IMO, you can just simply parse the whole "Address" key from the response. 

+ I also notice that in case of "cancelled" meetings, the "Address" value is usually updated with "canceled" text, so for such case, we just return empty location.

```js
// Multiple locations for "Address"
{
  "Date": "10/24/2024",
  "Type": "City Planning Commission",
  "APC": "Citywide",
  "CaseNumber": "",
  "BoardName": "",
  "Note": "CITY PLANNING COMMISSION\r\nREGULAR MEETING AGENDA",
  "Address": "21101 West Ventura Boulevard, 1515 South Veteran Avenue (1505, 1507, 1509, 1511, 1513, 1517, 1519, 1521, and 1523 South Veteran Avenue), 1666 North Vermont Avenue (1642 – 1666 North Vermont Avenue;\r\n4646 – 4650 West Prospect Avenue; and 4685 – 4697 West Hollywood Boulevard), ",
  "CD": "",
  "Agenda": "77621",
  "AgendaLink": "http://planning.lacity.gov/dcpapi2/meetings/document/77621",
  "Audio": "77677",
  "AudioLink": "http://planning.lacity.gov/dcpapi2/meetings/document/77677",
  "Minutes": "",
  "MinutesLink": "",
  "Hearing": "",
  "HearingLink": "",
  "AddDocs": "68278",
  "AddDocsLink": "http://planning.lacity.gov/dcpapi2/meetings/document/addtldoc/68278",
  "Time": "",
  "Description": "",
  "workshop": ""
}

// Canceled meeting has text "Canceled Meeting Agenda" included in "Address"
{
  "Date": "10/24/2024",
  "Type": "Area Planning Commission",
  "APC": "South Valley",
  "CaseNumber": "",
  "BoardName": "",
  "Note": "South Valley APC Canceled Meeting Agenda ",
  "Address": "South Valley APC Canceled Meeting Agenda ",
  "CD": "",
  "Agenda": "77622",
  "AgendaLink": "http://planning.lacity.gov/dcpapi2/meetings/document/77622",
  "Audio": "",
  "AudioLink": "",
  "Minutes": "",
  "MinutesLink": "",
  "Hearing": "",
  "HearingLink": "",
  "AddDocs": "",
  "AddDocsLink": "",
  "Time": "",
  "Description": "",
  "workshop": ""
}
```

LienDang reviewed (changes requested):
Please address the issues with 
```
start_urls
_parse_source()
_parse_links()
_parse_location()
```

msrezaie committed:
updated spider

- updated the logic of the following methods: start_urls, _parse_source(), _parse_links(), _parse_location()

msrezaie committed:
handled API rate limit

- fixed parse location conditional logic

msrezaie committed:
format fix

cruznunez commented on tests/test_losca_City_Planning.py:
```diff
@@ -0,0 +1,88 @@
+
+
+def test_status():
+    assert parsed_items[0]["status"] == "tentative"
```
<img width="435" src="https://github.com/user-attachments/assets/4397c7e9-ab10-4079-80e2-d23a6d4a453e" />
<img width="435" src="https://github.com/user-attachments/assets/c36252cb-4582-4fdd-b144-9092fc39aa60" />

I see some descriptions having "cancelled" in them. This should cause the meetings to have a status of cancelled. Adding some test cases for other statuses like cancelled and passed would be good.

cruznunez commented on city_scrapers/spiders/losca_City_Planning.py:
```diff
@@ -0,0 +1,106 @@
+                title=item.get("Type", ""),
+                description=item.get("Note", "").strip(),
+                classification=self._parse_classification(item),
+                start=parse(item.get("Date")),
```
<img width="435" alt="Screen Shot 2025-01-09 at 2 16 45 PM" src="https://github.com/user-attachments/assets/afddd891-8c83-4c4e-9a16-c0771383c53c" />

I see that all meetings get a start time of `00:00`. I see that the start times can be found in the linked PDFs. They seem to follow a pattern. Maybe we can have a default time per region? If there is not a pattern, maybe just add in the time notes to check the PDF for meeting time?

cruznunez reviewed (changes requested):
I think we can improve some tests and possibly the start times for the meetings and we should be good. Nice one!

msrezaie committed:
minor changes

- updated meeting time notes
- updated test file

cruznunez reviewed (approved):
LGTM

lamle-ea reviewed (approved):
Looks good to me. Once the CI task issue is resolved, this PR can be merged.

msrezaie committed:
Merge branch 'City-Bureau:main' into losca_City_Planning

msrezaie committed:
Update ci.yml

- adds a fix to CI check fail by cleaning the cache

msrezaie committed:
Update ci.yml

- removed py dependencies cache
- removed clean environment

lamle-ea committed:
Merge branch 'main' into losca_City_Planning
